### PR TITLE
Add donor role, nickname, and avatar integration with RW

### PR DESCRIPTION
**Pull request by rmcauley**
Draft status for now, please leave feedback and teach me how to test locally before pushing. :)

### CODE REVIEW

File: wormgas/cogs/rainwave.py [hunk 1]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
...
+        donors: List[str] = []
+        for guild in self.bot.guilds:
+            for member in guild.members:
+                if donor_role in member.roles:
+                    donors.append(member.id)
+        await self.rw_enable_perks(donors)
+
+    @cmds.command()
+    @cmds.is_owner()
+    async def sync_donors(self):
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:938]: `discord.errors.ClientException: Callback for sync_donors command is missing "ctx" parameter.`

```suggestion
    async def sync_donors(self, ctx: cmds.Context):
```

File: wormgas/cogs/rainwave.py [hunk 2]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
         else:
             await ctx.author.send('Your attempt to vote was not successful.')
 
+    @cmds.Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member):
+        if self.config.get("rainwave:donor_role_id"):
+            donor_role = self.config.get("rainwave:donor_role_id")
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:911]:
```suggestion
        if self.bot.config.get("rainwave:donor_role_id"):
            donor_role = self.bot.config.get("rainwave:donor_role_id")
```

File: wormgas/cogs/rainwave.py [hunk 3]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
...
+        if before.nick != after.nick:
+            await self.rw_update_nickname(after.id, after.nick)
+
+    @cmds.Cog.listener()
+    async def on_user_update(self, before: discord.User, after: discord.User):
+        if before.avatar != after.avatar:
+            await self.rw_update_avatar(after.id, after.avatar)
+
+    async def _sync_donors(self):
+        if not self.config.get("rainwave:donor_role_id"):
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:925]:
```suggestion
        if not self.bot.config.get("rainwave:donor_role_id"):
```

File: wormgas/cogs/rainwave.py [hunk 4]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
...
+            return
+
+        donor_role = self.config.get("rainwave:donor_role_id")
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:928]:
```suggestion
        donor_role = self.bot.config.get("rainwave:donor_role_id")
```

File: wormgas/cogs/rainwave.py [hunk 5]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
...
+        await self.rw_enable_perks(donors)
+
+    @cmds.command()
+    @cmds.is_owner()
+    async def sync_donors(self):
+        await self._sync_donors()
+
+    @cmds.Cog.listener()
+    async def on_ready(self):
+        if self.config.get("rainwave:sync_donor_role_on_ready"):
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:943]:
```suggestion
        if self.bot.config.get("rainwave:sync_donor_role_on_ready"):
```

File: wormgas/cogs/rainwave.py [hunk 6]
```diff
@@ -885,6 +905,44 @@ async def vote(self, ctx: cmds.Context, candidate: int):
...
+
+    async def _sync_donors(self):
+        if not self.config.get("rainwave:donor_role_id"):
+            return
+
+        donor_role = self.config.get("rainwave:donor_role_id")
+        donors: List[str] = []
+        for guild in self.bot.guilds:
+            for member in guild.members:
+                if donor_role in member.roles:
```
Comment (williamjacksn) [inline wormgas/cogs/rainwave.py:932]:
```suggestion
            role = guild.get_role(donor_role)
            for member in guild.members:
                if role in member.roles:
```

**Review by williamjacksn — CHANGES_REQUESTED** [review]
First time using GitHub's pull request review tools. Hopefully this works.

**Comment by rmcauley** [comment]
All fixes applied!

**Pull request merged**